daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/VariableMap1.scala

```diff
@@ -232,19 +345,33 @@ class VariableMap private(vTable: Map[GlobalQName, MStackOf[VariableInstance]])
    */
   def readVariable(vrd: VariableRuntimeData, referringContext: ThrowsSDE, maybePstate: Maybe[ParseOrUnparseState]): DataValuePrimitive = {
     val varQName = vrd.globalQName
-    val stack = vTable.get(varQName)
-    if (stack.isDefined) {
-      val variable = stack.get.top
-      variable.state match {
-        case VariableRead if (variable.value.isDefined) => variable.value.getNonNullable
-        case VariableDefined | VariableSet if (variable.value.isDefined) => {
+    val abuf = vTable.get(varQName)
+    if (abuf.isDefined) {
+      val variable = abuf.get.last
+      (variable.state, variable.value.isDefined, variable.defaultValueExpr.isDefined) match {
+        case (VariableRead, true, _) => variable.value.getNonNullable
+        case (VariableDefined | VariableSet, true, _) => {
           if (maybePstate.isDefined && maybePstate.get.isInstanceOf[PState])
```

Comment:
While we're renaming variables, can we change maybePstate to just maybeState, since it seems it can be either parse or unparse? I thought this was specifically for parse, and was confused about what happens during unparse, but that doesn't seem to be the case. Or should this always be PState, and maybe the type of maybePState should be Maybe[PState] in which case you only need th isDefined check?

Reply:
There is some book keeping that is specific to parsing only (see the call to markVariableRead) that doesn't apply to unparsing, hence the check.  I'll go ahead with the rename though.